lib/isolator.rb

```diff
@@ -69,27 +69,33 @@ def transactions_threshold
     end
 
     def transactions_threshold=(val)
+      # Do we need separate threshold per connection?
       Thread.current[:isolator_threshold] = val
     end
 
-    def incr_transactions!
-      Thread.current[:isolator_transactions] =
-        Thread.current.fetch(:isolator_transactions, 0) + 1
-      start! if Thread.current.fetch(:isolator_transactions) == transactions_threshold
+    def incr_transactions!(connection = ActiveRecord::Base.connection)
+      Thread.current[:isolator_connection_transactions] ||= {}
+      Thread.current[:isolator_connection_transactions][identifier_for(connection)] =
+        current_transactions(connection) + 1
+      start! if current_transactions(connection) == transactions_threshold
     end
 
-    def decr_transactions!
-      Thread.current[:isolator_transactions] =
-        Thread.current.fetch(:isolator_transactions) - 1
-      finish! if Thread.current.fetch(:isolator_transactions) == (transactions_threshold - 1)
+    def decr_transactions!(connection = ActiveRecord::Base.connection)
+      # Decrementing for unknown connection should raise error.
+      Thread.current[:isolator_connection_transactions][identifier_for(connection)] =
+        current_transactions(connection) - 1
+      finish! if current_transactions(connection) == (transactions_threshold - 1)
     end
 
     def clear_transactions!
-      Thread.current[:isolator_transactions] = 0
+      Thread.current[:isolator_connection_transactions] = {}
     end
 
     def within_transaction?
-      Thread.current.fetch(:isolator_transactions, 0) >= transactions_threshold
+      Thread.current.fetch(:isolator_connection_transactions, {}).each_value do |transaction_count|
+        return true if transaction_count >= transactions_threshold
```

Comment:
This isn't correct. we can't really use the same `transactions_threshold`. My thinking is the effective_threshold needs to be the sum of all non-zero connections' transaction_count

Reply:
I don't think sum would work.
I doesn't matter which connection has a transaction, we need  to raise even at least one violates its threshold.